lsコマンドオブジェクト指向版　修正

rira100000000 opened the pull request:
[lsコマンドを作るオブジェクト指向版 by rira100000000 · Pull Request \#12 · rira100000000/ruby\-practices](https://github.com/rira100000000/ruby-practices/pull/12)

こちらの指摘を修正したのち、クラス構成を変更したくなったためブランチを切りました。

rira100000000 committed:
lsコマンドにクラスを導入

lsコマンドにクラスを導入しました。

rira100000000 committed:
empty?メソッドを使用するように修正

配列の空判定でempty?メソッドを使用するように修正しました。

rira100000000 committed:
RuboCop対応

.rubocop.ymlにTargetRubyVersionを設定しました。

rira100000000 committed:
クラスの再実装

listクラスを削除し、separetorクラスを実装しました。
オブジェクト指向を意識した実装に修正しました。

rira100000000 committed:
empty?メソッドを使用するように修正

array == [] の条件式を array.empty?に修正しました

rira100000000 committed:
文字列連結にjoinを使用するように修正

連結したい文字列を配列にしてjoinするようにしました

rira100000000 committed:
{}が複数行の時の体裁修正

{}が複数行になる時の{}の位置を修正しました。

rira100000000 committed:
コメントの追加

rira100000000 committed:
long_formatの分岐処理をクラス内で行うように修正

make_long_format_listとmake_long_format_list_with_pathsの分岐を呼び出し側が意識しなくていいように修正しました。

rira100000000 committed:
不要な変数の削除

再利用しない変数を削除しました。

rira100000000 committed:
変数名とキー名の修正

変数名とキー名を正しいスペル、簡潔な名前に修正しました。

rira100000000 committed:
変数名の修正

rira100000000 committed:
条件式でnilではなくfalseを使うよう修正

true,falseを想定しているため、nilではなくfalseを使うように修正しました。

rira100000000 committed:
reverse!をrecerseに修正

破壊的メソッドを使用しないように修正しました。

rira100000000 committed:
display_lineへの代入処理のリファクタリング

ljstを使うように修正しました。

rira100000000 committed:
修正漏れの修正

キーワード引数にした際に呼び出し側を修正していなかったため、修正しました。

rira100000000 committed:
文字列の構築を配列とjoinで行うように修正

文字列の構築が読みづらかったため、配列とjoinを使うように修正しました。

rira100000000 committed:
仕様、設計の変更

パスの指定は1つまでにしました。
separatorクラスを削除しました。
FileListクラスを導入し、設計が以下のようになりました
* FileListクラスはファイルの詳細であるFileDetailを持っている
* LongFormatListクラス、ShortFormatListクラスはそれぞれFileListクラスを使ってフォーマットに沿ったリストを作る

rira100000000 committed:
reset_max_length_hashメソッドを削除

再利用しないのでinitializeに直接書くようにしました。

rira100000000 committed:
LongFormatListとShortFormatListの共通処理をモジュール化

共通の処理をFetchFileDetailモジュールにmake_formatted_listメソッドとして実装しました

rira100000000 committed:
FileDetailの各値をメソッド化

FileDetailの持つ各値をインスタンス変数化、メソッドで呼び出し可能にしました。

rira100000000 committed:
privateとpublicを分ける

各クラスのメソッドをprivateとpublicに分けました。

rira100000000 committed:
RuboCop対応

rira100000000 committed:
不要なインスタンス変数をローカル変数化

@pathは外部から呼び出さないのでローカル変数にしました。

rira100000000 committed:
AllCopsの削除

必要なくなったため、AllCopsを削除しました。

rira100000000 committed:
メソッド名、モジュール名の修正

メソッド名、モジュール名を適切なものに修正しました。

rira100000000 committed:
グローバルで定数を宣言しないよう修正

グローバルで定数を宣言していたため、クラス内で宣言するようにしました。

rira100000000 committed:
クラス設計の修正

LongFormatListとShortFormatListクラスはフォーマットだけを行うように修正しました。
ファイル情報の取得はFileDetailFetcherクラスで行うようにしました。

rira100000000 committed:
rオプションが正常に表示されない不具合修正

rira100000000 committed:
transposeを使うように修正

ファイル一覧を作成する際にtransposeメソッドを使うよう修正しました。

rira100000000 committed:
非ascii文字のずれを修正

非ascii文字を含むと列が乱れる不具合を修正しました。

rira100000000 committed:
テスト用ディレクトリの作成、

テスト用ディレクトリを作成し、.gitignoreに追加しました。

rira100000000 committed:
mapを使うように修正

each内で配列を作っていたのでmapを使うように修正しました。

rira100000000 committed:
代入方法を統一

一つのif文で別々の代入方法を使っていたので統一しました。

rira100000000 committed:
ループの引数名を修正

配列namesからfileを取り出していたので、nameを取り出すようにしました。

rira100000000 committed:
不要なインスタンス変数を削除

不要なインスタンス変数を削除して、呼び出し用にメソッドを追加しました。

rira100000000 committed:
不要なインスタンス変数を削除

不要なインスタンス変数を削除しました。

rira100000000 committed:
変数呼び出し用にメソッドを追加

uidとgidを呼び出すためのメソッドを追加しました。

rira100000000 committed:
FileDetailfetcherをls.rbでrequireする

long_format_listとshort_dormat_listでそれぞれrequireしていたので修正しました。

rira100000000 committed:
変数名を修正

rira100000000 committed:
mapを使うように修正

rira100000000 committed:
max_length_hashをStructに変更

ハッシュで定義していたmax_length_hashをStructに変更しました。

rira100000000 committed:
modeメソッドにget_file_modeの処理を移譲

rira100000000 committed:
FileDetailfetcherクラスの削除

FileDetailfetcherクラスを削除し、メソッドをls.rbに移しました。

rira100000000 committed:
不要なrequireと属性を削除しました

rira100000000 committed:
デメテルの法則に則るように修正

statから呼び出している属性をfile_detailクラスのメソッドとして呼び出せるようにしました。

rira100000000 committed:
file_list,rbを削除

file_list.rbを削除し、必要な処理はls.rbに移動しました

rira100000000 committed:
変数名の変更とコメントの追加

可読性を高めるため、変数名の変更とコメントを追加しました。

rira100000000 committed:
each_sliceを使ってリファクタリング

each_sliceを使って処理を書き直しました。

rira100000000 committed:
join前の配列を綺麗にする

join前の配列が[1,[2,3]]のような形になっていたので綺麗な1次配列になるようにしました。

rira100000000 committed:
.ruby-versionを追加

rira100000000 committed:
LongFormatListを`new`するタイミングでfile_detailsを渡すように修正

LongFormatListを`new`するタイミングでfile_detailsを渡すようにしました。
同様にShortFormatListでも`new`するタイミングでfile_detailsを渡すようにしました。

rira100000000 committed:
パスの連結にPathname.joinを使うように修正しました。

rira100000000 committed:
空文字列の判定にempty?を使うよう修正

空文字列の判定に=''を使用していたので、empty?に修正しました。

rira100000000 committed:
パスの指定機能を削除

必須要件出なかったのでパスの指定機能は削除しました

rira100000000 committed:
file_detailのinitializeにおける初期値の削除

initializeでdirectoryは必ず指定されるため、初期値を削除しました。

rira100000000 committed:
コマンドライン引数のパスを無視するよう修正

コマンドライン引数はa,l,rのみサポートするように修正しました。

rira100000000 committed:
formatメソッドをprivateに修正

外部から呼び出さないのでprivateに修正しました。

rira100000000 committed:
不要な変数の削除

calc_max_lengthをsplat展開し変数を作らないよう修正しました。

rira100000000 committed:
クラス構成の見直しと修正

long_format_listクラスをlong_list_formatterモジュールに変更しました。
short_format_listクラスをshort_list_formatterモジュールに変更しました。
file_detailの集合を扱うfile_listクラスを追加しました。

JunichiIto commented on .gitignore:
```diff
@@ -50,7 +51,7 @@ build-iPhoneSimulator/
 # intended to run in multiple environments; otherwise, check them in:
 # Gemfile.lock
-# .ruby-version
+.ruby-version
```
`.ruby-version`を有効にしたのはなぜなんでしょう？

rira100000000 replied:
ローカルのRubyのバージョンが3.1だったことに気づいて3.2に変更したのですが、その際に`.ruby-version`が生成されました。`.ruby-version`をコミットするべきかわからなかったので調べたところ、こちらの記事を見つけました。
[\`bundle gem\` で gem を作成する時の \.gitignore について \- Qiita](https://qiita.com/sekizo/items/3b4e90145461f1e38db4#ruby-version)

ですがよく考えるとgemと単体で動作するプログラムでは理屈が異なりますね…
改めて調べたところ`.ruby-version`はgitの管理対象にすべきということがわかりましたので`.gitignore`から除外します。

JunichiIto replied:
情報ありがとうございます！
`.ruby-version`をgitに含めるかどうかはケースバイケースですね。
Railsアプリだと、どちらかというと含めることが多いような気がしますが。

ただ、どちらにしてもこのプラクティスには直接関係ないのでdiffには含めない方がいいですね。

rira100000000 replied:
なるほどです。プラクティスと関係あるか？という観点が抜けてました。

`rbenv shell 3.2.1`で問題なく動かせましたので`.ruby-version`自体削除します。

JunichiIto commented on 07.ls_object/file_detail.rb:
```diff
@@ -0,0 +1,72 @@
+# frozen_string_literal: true
+
+class FileDetail
+  require 'pathname'
```
class構文の外でrequireすることが多いですね。

rira100000000 replied:
class構文の外でrequireします。

JunichiIto commented on 07.ls_object/file_detail.rb:
```diff
@@ -0,0 +1,72 @@
+  end
+
+  def mode
+    # ファイルモードを8進数に変換して末尾3桁（パーミッション）を取得する
```
WHYの説明にはなってないですね。

[【動画付き・初心者向け】コメントにはWHYを書こう 〜不要なコメントと必要なコメントの違いについて〜 \- Qiita](https://qiita.com/jnchito/items/f0d90af4ed44b7484103)

rira100000000 replied:
コメントの内容を「8進数に変換したファイルモードの末尾3桁はパーミッションを表している」に修正します。

JunichiIto commented on 07.ls_object/file_list.rb:
```diff
@@ -0,0 +1,21 @@
+    @file_details = sorted_file_names.map do |name|
+      FileDetail.new(name, directory)
+    end
+  end
```
initializeしかないクラスというのは不自然ですね。
クラスの実体はLongListFormatterとShortListFormatterだと思われるので、これらをクラスにした方がいいです。

ファイルをかき集めてくる処理はクラスの外で実行して、それをFomatterクラスに渡すような設計にしてみましょう。
（ほぼ同じコメントを以前も書いてました→ https://github.com/rira100000000/ruby-practices/pull/12#discussion_r1185625747 ）

モジュールに分けたとしてもクラスを利用する側からすると、モジュールを使ってるかどうかは関係ない（外から見えるのはあくまでクラスだけ）なので、結局「1つで何でもできちゃうクラス」になってます。

rira100000000 replied:
確かにこれでは名前の違う`FileDetailfetcher`を作っているだけでした🙇‍♂️
LongListFormatterとShortListFormatterをクラスとします。

JunichiIto commented on 07.ls_object/long_list_formatter.rb:
```diff
@@ -0,0 +1,52 @@
+
+module LongListFormatter
+  def long_format
+    max_length = Data.define(:nlink, :uid, :gid, :file_size, :file_name).new(* calc_max_length)
```
```suggestion
    max_length = Data.define(:nlink, :uid, :gid, :file_size, :file_name).new(*calc_max_length)
```

スペースを空けない方が多いかも。

あと、ここの`:nlink, :uid, :gid, :file_size, :file_name`と、50行目の`[nlink, uid, gid, file_size, file_name]`をリンクさせる（カラムの順番を一致させる）のは「もろい」プログラムになる原因になります。

https://bootcamp.fjord.jp/pages/three-legged-race#-3

```suggestion
    max_length = hogehoge
```

のようにオブジェクトの作成そのものをメソッド化した方が堅牢になりそうです。

rira100000000 replied:
`calc_max_length`でオブジェクトの作成までするようにします。

`*`はスペースを空けないようにします。

JunichiIto commented on 07.ls_object/long_list_formatter.rb:
```diff
@@ -0,0 +1,52 @@
+
+  def calc_total_block
+    # statの1ブロック単位は512byte
+    # lsコマンドでの1ブロック単位1024byteに合わせるため2で割る
```
こういうのはWHYの説明になってていいですね

JunichiIto commented on 07.ls_object/short_list_formatter.rb:
```diff
@@ -0,0 +1,45 @@
+    file_name.each_char.sum do |char|
+      char.ascii_only? ? 0 : 1
+    end
+  end
```
`calc_file_name_size`とロジックがほぼ同じなので、

```ruby
  def calc_file_name_size(file_name)
    hogehoge(1)
  end

  def count_not_ascii(file_name)
    hogehoge(0)
  end
```

みたいにDRYできるといいかも

rira100000000 replied:
以下のように修正しました。
```ruby
  def calc_file_name_size(file_name)
    calc_ascii(file_name, 1)
  end

  def count_not_ascii(file_name)
    calc_ascii(file_name, 0)
  end

  def calc_ascii(file_name, ascii_size)
    file_name.each_char.sum do |char|
      char.ascii_only? ? ascii_size : ascii_size + 1
    end
  end
```

例示に従い`calc_file_name_size`と`count_not_ascii`は残したのですが、『メソッドを呼び出しているだけの不要なメソッド』に該当するのでは？と思っています。
ファイル名のサイズ計算と非ascii文字の合計では`calc_ascii`の用途が異なるので、読み手の理解を助けるためにそれぞれメソッドとして存在しても良いという事でしょうか？

JunichiIto replied:
good questionですね！
『メソッドを呼び出しているだけの不要なメソッド』に該当する、ように見えるかもしれませんが、独自の引数を追加しているので厳密に言えば該当しないですね。たとえば以下のようなコードになっていれば該当すると思います。

```ruby
  def calc_file_name_size(file_name)
    calc_ascii(file_name)
  end

  def count_not_ascii(file_name)
    calc_ascii(file_name)
  end
```

あと、「読み手の理解を助けるため」という役割もありそうなので、ここはメソッドを残したままでも良いように思います。

rira100000000 replied:
なるほどです！
修正せずそのまま残しておきます！

JunichiIto commented on 07.ls_object/ls.rb:
```diff
@@ -0,0 +1,30 @@
+
+  file_list = FileList.new(options[:r], options[:a])
+  if options[:l]
+    puts file_list.long_format
```
現状の実装でも以下でコメントした問題（「削除したファイルがまだ表示される」の件）は発生しそうです。

https://github.com/rira100000000/ruby-practices/pull/12#discussion_r1190472296

rira100000000 replied:
この問題は『最新の情報を取得する処理なのに、保持した情報を利用している』ことが原因だと考えました。
以上を踏まえて以下のように修正しました。

```ruby
def main
  options = parse_option

  flag = options[:a] ? File::FNM_DOTMATCH : 0
  directory = Dir.pwd
  names = Dir.glob('*', base: directory, flags: flag).sort
  sorted_file_names = options[:r] ? names.reverse : names

  if options[:l]
    # file_detailsは作成したら保持しないで即利用する
    puts LongListFormatter.new.format(fetch_file_details(sorted_file_names, directory))
  else
    puts ShortListFormatter.new.format(fetch_file_details(sorted_file_names, directory))
  end
end

def fetch_file_details(file_names, directory)
  file_names.map do |name|
    FileDetail.new(name, directory)
  end
end
```

ここで疑問なのですが、LongListFormatterやShortListFormatterは何もデータを保持しませんが、インスタンスを生成する必要があるのでしょうか？
`format`をクラスメソッドとして呼び出せれば十分では？と思うのですが、そうするとクラスメソッドしか持たないクラスができてしまい、それはそれで不自然に感じます。
現状だと『クラスメソッドしか持たないクラスは不自然だから』という理由でインスタンスメソッドとして定義してしまっているので、理解が足りていないと感じています。

JunichiIto replied:
> ここで疑問なのですが、LongListFormatterやShortListFormatterは何もデータを保持しませんが、インスタンスを生成する必要があるのでしょうか？

これもgood questionですね。
「クラスメソッドしか持たないクラス」はRubyの場合モジュールにしてしまう、という手もあります。
（チェリー本8.7参照）

あと、`LongListFormatter.new(file_details).format` のように、newでデータを渡すのが絶対に間違いというわけではないです。これは設計の方針や開発者の好みによって、こっちを採用する人も出てくると思います。

また、今のコードのまま（`LongListFormatter.new.format(file_details)` ）でもやはり「絶対にダメ」とは言い切れないです。
というのも、「クラスらしくないクラス」が多少出てくること自体はオブジェクト指向的に間違いではないからです。
（ https://bootcamp.fjord.jp/products/15703#comment_129486 でも「決して作ってはいけないわけではない。こんなクラスばかりになることが問題」と書いています）

なので、いずれかriraさんの好きな選択肢を選んでもらう、って感じですかね〜。

オブジェクト指向プログラミングは（というか、OOPに限らずですが）、「明らかにおかしい、と言えるコードはあっても、これが絶対に正しい、と言い切れるコードは滅多にない」というのがなかなか難しく、かつ面白いところでもあります。

rira100000000 replied:
>「明らかにおかしい、と言えるコードはあっても、これが絶対に正しい、と言い切れるコードは滅多にない」

これ面白いですね～👀

モジュールにするかクラスにするかですが、ダックタイピング的に`format`メソッドを呼びたいので、シンプルに書けるクラスで実装したいと思います。
```ruby
formatter = if options[:l]
                LongListFormatter.new
              else
                ShortListFormatter.new
              end

puts formatter.format(create_file_details(sorted_file_names))
```

JunichiIto reviewed (commented):
コメントしました！

rira100000000 committed:
コメントの修正

WHYを説明する文に修正しました。

rira100000000 committed:
.ruby-versionをgit管理対象に追加

/ruby-versionを.gitignoreから除外しました。

rira100000000 committed:
requireをクラス構文の外で行うよう修正

rira100000000 committed:
オブジェクトの作成をメソッドで行うように修正

rira100000000 committed:
calc_file_name_sizeとcount_not_asciiをDRYにする

calc_file_name_sizeとcount_not_asciiで共通する処理をメソッドにしました。

rira100000000 committed:
Fomatterクラスの作成

LongListFormatterとShortListFormatterクラスを作成しました。
ファイルの取得処理と成形処理を分離しました。

JunichiIto commented on 07.ls_object/.ruby-version:
```diff
@@ -0,0 +1 @@
+3.2.1
```
悩ましいところですが、バージョンの指定はプラクティスとは直接関係ないのでdiffに入れない方がいいでしょうね。
`rbenv shell 3.2.1`のような形で`.ruby-version`を作成しない方式にした方がベターです。

とはいえ、「VS Codeで動かしたりするのにどうしても必要」みたいな理由であれば作成しないと不便だと思うので、「やむをえない」理由があればOKにしようと思います。

rira100000000 replied:
`rbenv shell 3.2.1`で問題なく動かせましたので`.ruby-version`は削除します。

JunichiIto commented on 07.ls_object/ls.rb:
```diff
@@ -16,14 +18,24 @@ def parse_option
+    puts LongListFormatter.new.format(fetch_file_details(sorted_file_names, directory))
   else
-    puts file_list.short_format
+    puts ShortListFormatter.new.format(fetch_file_details(sorted_file_names, directory))
```
`fetch_file_details(sorted_file_names, directory)`はDRYにできそう

rira100000000 replied:
以下のように修正します。
```ruby
formatter = if options[:l]
                LongListFormatter.new
              else
                ShortListFormatter.new
              end

puts formatter.format(create_file_details(sorted_file_names))
```

JunichiIto commented on 07.ls_object/ls.rb:
```diff
@@ -16,14 +18,24 @@ def parse_option
   options
 end
 
+def fetch_file_details(file_names, directory)
```
fetchは「OSのファイルシステムやWeb APIなど、プログラムの外からデータを取ってくる」というときによく使います。
ここはfile_nameをFileDetailに変換しているだけのように見えますね。

rira100000000 replied:
`FileDetail.new(path)`することで`path`から`stat`を取得するのでプログラムの外から情報を取ってくる、つまりfetchだ！と考えたのですが、改めて考えると`FileDetail.new`の本質は`fetch`ではないですね🤔

`create_file_details`に変更したいと思います。

JunichiIto commented on 07.ls_object/ls.rb:
```diff
@@ -16,14 +18,24 @@ def parse_option
 
+def fetch_file_details(file_names, directory)
+  file_names.map do |name|
+    FileDetail.new(name, directory)
```
newの引数はnameとdirectoryにわけずに、pathをまるっと渡す方がAPIとしてはシンプルな気がしました。

rira100000000 replied:
FileDetailの`initialize`で`path`を受け取るように修正します。

JunichiIto commented on 07.ls_object/ls.rb:
```diff
@@ -16,14 +18,24 @@ def parse_option
 
-  file_list = FileList.new(options[:r], options[:a])
+  flag = options[:a] ? File::FNM_DOTMATCH : 0
+  directory = Dir.pwd
+  names = Dir.glob('*', base: directory, flags: flag).sort
```
```suggestion
  names = Dir.glob('*', flags: flag).sort
```

でも同じ動きになりませんかねー？

rira100000000 replied:
なりました！
デフォルトでカレントディレクトリを対象にしていたので`base`は削除します。

JunichiIto commented on 07.ls_object/file_detail.rb:
```diff
@@ -40,7 +40,7 @@ def type
   def mode
-    # ファイルモードを8進数に変換して末尾3桁（パーミッション）を取得する
+    # 8進数に変換したファイルモードの末尾3桁はパーミッションを表している
     @stat.mode.to_s(8)[-3..].each_char.map do |char|
```
```suggestion
    permission = @stat.mode.to_s(8)[-3..]
    permission.each_char.map do |char|
```

みたいにすればコメントが不要になるかも（説明的変数の導入）

rira100000000 replied:
コメントを削除して説明的変数を導入します。

JunichiIto reviewed (commented):
コメントしました！

rira100000000 committed:
mainをDRYにする
fetch_file_detailsを呼び出す部分DRYにしました。

rira100000000 committed:
.ruby-versionを削除しました

rira100000000 committed:
globでbaseの指定部分を削除

globはデフォルトでカレントディレクトリを対象とするのでbaseの記述は削除しました。

rira100000000 committed:
FileDetailのinitializeではpathのみを受け取るように修正しました

rira100000000 committed:
fetch_file_detailsをcreate_file_detailsにリネームしました。

rira100000000 committed:
説明的変数の導入

コメントを削除し、説明的変数を導入しました。

JunichiIto commented on 07.ls_object/file_detail.rb:
```diff
@@ -28,10 +28,8 @@ class FileDetail
-
-    path = Pathname.new(directory).join(name).to_s
+  def initialize(path)
+    @name = Pathname.new(path).basename.to_s
```
文字列を返すのでto_sは不要ですね。
https://docs.ruby-lang.org/ja/latest/method/File/s/basename.html

rira100000000 replied:
~`Pathname`クラスの`basename`だと`Pathname`型が返ってきています🙏~

[Pathname\#basename \(Ruby 3\.2 リファレンスマニュアル\)](https://docs.ruby-lang.org/ja/latest/method/Pathname/i/basename.html)

rira100000000 replied:
newの際にファイル名を直接指定できたので、`basename`メソッドは不要になりました。
```ruby
FileDetail.new(name)
```

JunichiIto commented on 07.ls_object/ls.rb:
```diff
@@ -18,25 +18,27 @@ def parse_option
+def create_file_details(file_names)
   file_names.map do |name|
-    FileDetail.new(name, directory)
+    path = Pathname.new(Dir.pwd).join(name).to_s
+    FileDetail.new(path)
```
```suggestion
    FileDetail.new(name)
```

でも問題なく動いたりしませんか？（Rubyのファイル関連のAPIはデフォルトでカレントディレクトリを起点にして動くことが多いような）

rira100000000 replied:
問題なく動きました！
ファイル名を直接渡すようにします。

JunichiIto commented on 07.ls_object/ls.rb:
```diff
@@ -18,25 +18,27 @@ def parse_option
+                LongListFormatter.new
+              else
+                ShortListFormatter.new
+              end
```
好みですが三項演算子でもいいかもしれません。

rira100000000 replied:
スッキリかけるので三項演算子に修正します。

JunichiIto commented on 07.ls_object/ls.rb:
```diff
@@ -18,25 +18,27 @@ def parse_option
+                ShortListFormatter.new
+              end
+
+  puts formatter.format(create_file_details(sorted_file_names))
```
ポリモーフィズムを活用して良い感じです 👍

JunichiIto replied:
好みと言えば好みなのですが、僕は

```ruby
options = parse_options
file_details = create_file_details(options)
formatter = options[:l] ? LongListFormatter.new : ShortListFormatter.new
puts formatter.format(file_details)
```

みたいになってる方が読みやすいな〜と思います。（コードの各行が以下のように文章の見出しっぽく並ぶイメージです）

```ruby
# オプションを取得
# ファイル情報を取得
# フォーマッターを取得
# 整形して出力
```

rira100000000 replied:
確かにこちらの方が読みやすそうです。
例示に従って修正します。

JunichiIto reviewed (commented):
コメントしました！

rira100000000 committed:
FileDetail.newの引数をpathからnameに変更しました

rira100000000 committed:
条件式を三項演算子に変更しました。

rira100000000 committed:
不要なrequireを削除しました。

JunichiIto commented on 07.ls_object/ls.rb:
```diff
@@ -0,0 +1,39 @@
+# frozen_string_literal: true
+
+require 'optparse'
+require 'etc'
```
標準ライブラリのrequireは、それを必要としているファイルで都度requireすることが多い気がします。

rira100000000 replied:
`require 'etc'`をfile_detail.rbに移動します。

JunichiIto commented on 07.ls_object/ls.rb:
```diff
@@ -0,0 +1,39 @@
+end
+
+def main
+  options = parse_option
```
単数か複数か揃えたいなと思いました。

rira100000000 replied:
`parse_option`を複数形にします。

JunichiIto commented on 07.ls_object/short_list_formatter.rb:
```diff
@@ -0,0 +1,47 @@
+  SPACE_FOR_COLUMNS = 2
+
+  def format(file_details)
+    last_row = (file_details.size.to_f / COLUMNS).ceil
```
最後の行データ？と読めなくもないので、`row_count`（行数）みたいな名前の方がいいかもしれません。

rira100000000 replied:
`row_count`に修正します。

JunichiIto commented on 07.ls_object/short_list_formatter.rb:
```diff
@@ -0,0 +1,47 @@
+
+  private
+
+  def add_space_for_line(file_names_list, max_file_names)
```
メソッド名通り受け取ると、「スペースを追加するだけのメソッドかな？」と思いますが、実際にはそれ以外にもあれこれやってるので、より適切な名前が求められる気がします。

rira100000000 replied:
`long_list_formatter`に合わせて`adjust_list_for_display`に変更します。

JunichiIto reviewed (commented):
コメントしました〜。

rira100000000 committed:
標準ライブラリのrequireを必要としているファイルで行うよう修正

標準ライブラリのrequireは、それを必要としているファイルで都度requireするようにしました。

rira100000000 committed:
複数形と単数形を統一する

options = parse_optionが複数形と単数形で不一致だったので、複数形に統一しました。

rira100000000 committed:
変数名の見直し

last_rowをrow_countに修正しました。

rira100000000 committed:
メソッド名の修正

add_space_for_lineをより実装に則した名前に修正しました。

rira100000000 committed:
mainのリファクタリング

可読性を高めるため、file_details変数を作りました。
create_file_detailsにoptionsを渡し、create_file_details内でoptionsの処理をするようにしました。

JunichiIto commented on 07.ls_object/short_list_formatter.rb:
```diff
@@ -5,23 +5,23 @@ class ShortListFormatter
   def format(file_details)
-    last_row = (file_details.size.to_f / COLUMNS).ceil
+    row_count = (file_details.size.to_f / COLUMNS).ceil
     file_names_list = []
```
listには「配列」の意味で使われることがあります。file_namesも複数形なので配列を表します。
なので「`file_names_list` = ファイル名配列の配列」みたいな名前を付けていることになります。

ここは`[[1,2,3],[4,5,6]]`のようにネストした配列になるので、たとえば`nested_file_names`みたいな名前にする方法があります。

もしくは

```
[
  [1,2,3],
  [4,5,6]
]
```

のように配列で行と列で構成されるテーブルを構成している、と考えれば`file_name_table`という名前を付けるのもアリです。

JunichiIto reviewed (approved):
1点コメントしましたが大きな問題ではないのでこれでOKです！🙆‍♂️
